redbot/core/utils/views.py

```diff
@@ -431,3 +431,147 @@ async def auth_button(self, interaction: discord.Interaction, button: discord.Bu
+        self.disable_buttons = disable_buttons
+
+    async def on_timeout(self):
+        if self.message is None:
+            # we can't do anything here if message is none
+            return
+
+        if self.disable_buttons:
+            self.confirm_button.disabled = True
+            self.dismiss_button.disabled = True
+            await self.message.edit(view=self)
+        else:
+            await self.message.edit(view=None)
+
+    @discord.ui.button(label=_("Yes"), style=discord.ButtonStyle.green)
+    async def confirm_button(self, interaction: discord.Interaction, button: discord.ui.Button):
+        """
+        A ``discord.ui.Button`` to confirm the message.
+
+        Examples
+        --------
+        Changing the style and label of this ``discord.ui.Button``::
+
+            view = ConfirmView(ctx.author)
+            view.confirm_button.style = discord.ButtonStyle.red
+            view.confirm_button.label = "Delete"
+            view.dismiss_button.label = "Cancel"
+            view.message = await ctx.send("Are you sure you want to remove #very-important-channel?", view=view)
+            await view.wait()
+            if view.result:
+                await ctx.send("Channel #very-important-channel deleted.")
+            else:
+                await ctx.send("Canceled.")
+        """
+        self.result = True
+        self.stop()
+        # respond to the interaction so the user does not see "interaction failed".
+        await interaction.response.defer()
+        # call `on_timeout` explicitly here since it's not called when `stop()` is called.
+        await self.on_timeout()
+
+    @discord.ui.button(label=_("No"), style=discord.ButtonStyle.secondary)
+    async def dismiss_button(self, interaction: discord.Interaction, button: discord.ui.Button):
+        """
+        A ``discord.ui.Button`` to dismiss the message.
+
+        Examples
+        --------
+        Changing the style and label of this ``discord.ui.Button``::
+
+            view = ConfirmView(ctx.author)
+            view.confirm_button.style = discord.ButtonStyle.red
+            view.confirm_button.label = "Delete"
+            view.dismiss_button.label = "Cancel"
+            view.message = await ctx.send("Are you sure you want to remove #very-important-channel?", view=view)
+            await view.wait()
+            if view.result:
+                await ctx.send("Channel #very-important-channel deleted.")
+            else:
+                await ctx.send("Canceled.")
+        """
+        self.result = False
+        self.stop()
+        # respond to the interaction so the user does not see "interaction failed".
+        await interaction.response.defer()
+        # call `on_timeout` explicitly here since it's not called when `stop()` is called.
+        await self.on_timeout()
+
+    async def interaction_check(self, interaction: discord.Interaction):
+        if self.message is None:
+            self.message = interaction.message
```

Comment:
Looking at the documentation, it seems that it's consistent with it but I do wonder if we actually want to make the behavior of the view's timeout action dependent on whether some random person other than the author decides to click one of the buttons under the message? Would it not make more sense to this API's consumers to expect that not setting `view.message` will fully prevent it from removing/disabling the buttons on the message?

Reply:
I suppose my original thinking was to prefer cleanup of the view in as many situations as possible. For example, a cog sends the view without explicitly setting the message attribute, if the issuer of the command doesn't click the button but someone else does it would be better to then cleanup the view even if it doesn't actually respond to them. I view it as better in that the buttons no longer sit around unusable for anyone. I could move it under the author check to be more consistent and keep everything else the same I suppose. Though now I also wonder if I should be responding to someone who isn't the author with a message as well.

Reply:
> Though now I also wonder if I should be responding to someone who isn't the author with a message as well.

I haven't thought about it but it would definitely make it seem not broken to bystanders. And it would be more consistent with all the other views/modals that we already have (just the simple menu and set api so far).

> I could move it under the author check to be more consistent and keep everything else the same I suppose.

It seems like an alright way of handling it. But with your explanation, I honestly am not sure if we should do that or just leave it as is (since I guess that the option of not trying to do the cleanup at all is not something we're interested in).

Reply:
I added a response if someone other than the author attempts to interact with the buttons. I left in setting the message object from any interaction though. I am much happier with the way this would work now and I would rather prioritize removing or disabling the view in as many situations as possible. Leaving broken buttons lying around outside of development and testing just feels bad.